python/semantic_kernel/planning/plan.py

```diff
@@ -1,17 +1,432 @@
+
+    @property
+    def next_step_index(self) -> int:
+        return self._next_step_index
+
+    def __init__(
+        self,
+        name: Optional[str] = None,
+        skill_name: Optional[str] = None,
+        description: Optional[str] = None,
+        next_step_index: Optional[int] = None,
+        state: Optional[ContextVariables] = None,
+        parameters: Optional[ContextVariables] = None,
+        outputs: Optional[List[str]] = None,
+        steps: Optional[List["Plan"]] = None,
+        function: Optional[SKFunctionBase] = None,
+    ) -> None:
+        self._name = "" if name is None else name
+        self._skill_name = "" if skill_name is None else skill_name
+        self._description = "" if description is None else description
+        self._next_step_index = 0 if next_step_index is None else next_step_index
+        self._state = ContextVariables() if state is None else state
+        self._parameters = ContextVariables() if parameters is None else parameters
+        self._outputs = [] if outputs is None else outputs
+        self._steps = [] if steps is None else steps
+        self._has_next_step = len(self._steps) > 0
+        self._is_semantic = None
+        self._function = None if function is None else function
+        self._request_settings = None
+
+        if function is not None:
+            self.set_function(function)
+
+    async def invoke_async(
+        self,
+        input: Optional[str] = None,
+        context: Optional[SKContext] = None,
+        settings: Optional[CompleteRequestSettings] = None,
+        memory: Optional[SemanticTextMemoryBase] = None,
+        logger: Optional[Logger] = None,
+        # TODO: cancellation_token: CancellationToken,
+    ) -> SKContext:
+        if input is not None:
+            self._state.update(input)
+
+        if context is None:
+            context = SKContext(
+                variables=self._state,
+                skill_collection=None,
+                memory=memory,
+                logger=logger,
+            )
+
+        if self._function is not None:
+            result = await self._function.invoke_async(
+                context=context, settings=settings
+            )
+            if result.error_occurred:
+                result.log.error(
+                    msg="Something went wrong in plan step {0}.{1}:'{2}'".format(
+                        self._skill_name, self._name, context.last_error_description
+                    )
+                )
+                return result
+            context.variables.update(result.result)
+        else:
+            # loop through steps until completion
+            while self.has_next_step:
+                function_context = context
+                self.add_variables_to_context(self._state, function_context)
+                await self.invoke_next_step(function_context)
+                self.update_context_with_outputs(context)
+
+        return context
+
+    def invoke(
```

Comment:
general question, can `invoke` call and wrap `invoke_async`, or is that considered bad practice? I'm just a little worried about having dual implementations of all the context variable handling, in case they get out of alignment. Good opportunity for some refactoring. (future PR++)

Reply:
I noticed it to. I think we should do this yes (in a future PR)